src/Juvix/Compiler/Backend/Geb/Translation/FromCore.hs

```diff
@@ -237,28 +254,64 @@ fromCore tab = case tab ^. Core.infoMain of
         unsupported
 
     convertBinop :: Opcode -> [Core.Node] -> Trans Morphism
-    convertBinop op args =
-      case args of
-        [arg1, arg2] -> do
-          arg1' <- convertNode arg1
-          arg2' <- convertNode arg2
-          return $ MorphismBinop (Binop op arg1' arg2')
-        _ ->
-          error "wrong builtin application argument number"
+    convertBinop op = \case
+      [arg1, arg2] -> do
+        arg1' <- convertNode arg1
+        arg2' <- convertNode arg2
+        return $
+          MorphismBinop
+            Binop
+              { _binopOpcode = op,
+                _binopLeft = arg1',
+                _binopRight = arg2'
+              }
+      _ ->
+        error "wrong builtin application argument number"
 
     convertConstr :: Core.Constr -> Trans Morphism
     convertConstr Core.Constr {..} = do
-      prod <- mkProd
-      return (foldr ($) prod (replicate tagNum MorphismRight))
+      args <- convertProduct _constrArgs
+      unless (tagNum < length constructors) $
+        error "constructor tag out of range"
```

Comment:
This can never be false unless there is a bug in `mkConstructors` or `convertInductive`.